File: solutions/2870.java
class Solution {
    public int minOperations(int[] nums) {
        int min_operation = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < nums.length; i++){
            if(map.containsKey(nums[i])){
                map.put(nums[i], map.get(nums[i]) + 1);
            }
            else { map.put(nums[i], 1); }
        }
        for(Integer value : map.values()){
            if(value == 1){
                return -1;
            }
            min_operation += value / 3;
            if(value % 3 != 0){
                min_operation++;
            }
        }
        return min_operation;
    }
}

File: solutions/2870. Minimum Number of Operations to Make Array Empty/2870.cpp
class Solution {
 public:
  int minOperations(vector<int>& nums) {
    int ans = 0;
    unordered_map<int, int> count;

    for (const int num : nums)
      ++count[num];

    for (const auto& [_, freq] : count) {
      // If freq == 3k, need k operations.
      // If freq == 3k + 1 = 3*(k - 1) + 2*2, need k + 1 operations.
      // If freq == 3k + 2, need k + 1 operations.
      if (freq == 1)
        return -1;
      ans += (freq + 2) / 3;
    }

    return ans;
  }
};
